File: ffmpeg-audio-encode/src/main/java/com/onzhou/recorder/GLProgram.java
package com.onzhou.recorder;

import android.opengl.GLES20;

public class GLProgram {

    private static final String TAG = GLProgram.class.getSimpleName();

    private int vertexShader = 0;
    private int fragmentShader = 0;
    private int program = 0;

    public int init(String vertexCode, String fragmentCode) {
        int vertexShader = GLES20.glCreateShader(GLES20.GL_VERTEX_SHADER);
        if (vertexShader > 0) {
            GLES20.glShaderSource(vertexShader, vertexCode);
            GLES20.glCompileShader(vertexShader);
            int[] compile = new int[1];
            GLES20.glGetShaderiv(vertexShader, GLES20.GL_COMPILE_STATUS, compile, 0);
            if (compile[0] != GLES20.GL_TRUE) {
                LogUtil.e(TAG, "Compile vertex shader failed.");
                GLES20.glDeleteShader(vertexShader);
                this.vertexShader = 0;
                return -1;
            }
            this.vertexShader = vertexShader;
        }

        int fragmentShader = GLES20.glCreateShader(GLES20.GL_FRAGMENT_SHADER);
        if (fragmentShader > 0) {
            GLES20.glShaderSource(fragmentShader, fragmentCode);
            GLES20.glCompileShader(fragmentShader);
            int[] compile = new int[1];
            GLES20.glGetShaderiv(fragmentShader, GLES20.GL_COMPILE_STATUS, compile, 0);
            if (compile[0] != GLES20.GL_TRUE) {
                LogUtil.e(TAG, "Compile fragment shader failed.");
                GLES20.glDeleteShader(fragmentShader);
                this.fragmentShader = 0;
                return -1;
            }
            this.fragmentShader = fragmentShader;
        }

        int program = GLES20.glCreateProgram();
        if (program == 0) {
            LogUtil.e(TAG, "Create program failed.");
            return -1;
        }
        GLES20.glAttachShader(program, this.vertexShader);
        GLES20.glAttachShader(program, this.fragmentShader);
        GLES20.glLinkProgram(program);
        int[] linkStatus = new int[1];
        GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, linkStatus, 0);
        if (linkStatus[0] != GLES20.GL_TRUE) {
            LogUtil.e(TAG, "Link program failed.");
            GLES20.glDeleteProgram(program);
            this.program = 0;
            return -1;
        }
        this.program = program;

        return 0;
    }

    public void use() {
        if (program > 0) {
            GLES20.glUseProgram(program);
        } else {
            LogUtil.e(TAG, "program has not been initialized.");
        }
    }

    public void unuse() {
        GLES20.glUseProgram(0);
    }

    public int getProgramID() {
        return program;
    }

    public void release() {
        if (vertexShader > 0) {
            GLES20.glDeleteShader(vertexShader);
        }
        if (fragmentShader > 0) {
            GLES20.glDeleteShader(fragmentShader);
        }
        if (program > 0) {
            GLES20.glDeleteProgram(program);
        }
    }

}

File: ffmpeg-audio-encode/src/main/cpp/GLProgram.cpp
//
// Created by xiangweixin on 2020/6/28.
//

#include "GLProgram.h"

int GLProgram::init(const char *vertexCode, const char *fragmentCode) {
    m_iVertexShader = glCreateShader(GL_VERTEX_SHADER);
    if (m_iVertexShader > 0) {
        glShaderSource(m_iVertexShader, 1, &vertexCode, nullptr);
        glCompileShader(m_iVertexShader);
        GLint status = GL_FALSE;
        glGetShaderiv(m_iVertexShader, GL_COMPILE_STATUS, &status);
        if (status != GL_TRUE) {
            GLchar log[1024];
            GLsizei length;
            glGetShaderInfoLog(m_iVertexShader, 1024, &length, log);
            LOGE("glCompile vertex shader failed. ret is %d, info is %s", status, log);
            release();
            return GL_COMPILE_ERROR;
        }
    }

    m_iFragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
    if (m_iFragmentShader > 0) {
        glShaderSource(m_iFragmentShader, 1, &fragmentCode, nullptr);
        glCompileShader(m_iFragmentShader);
        GLint status = GL_FALSE;
        glGetShaderiv(m_iFragmentShader, GL_COMPILE_STATUS, &status);
        if (status != GL_TRUE) {
            GLchar log[1024];
            GLsizei length;
            glGetShaderInfoLog(m_iFragmentShader, 1024, &length, log);
            LOGE("glCompile fragment shader failed. ret is %d, info is %s", status, log);
            release();
            return GL_COMPILE_ERROR;
        }
    }

    m_iProgram = glCreateProgram();
    if (m_iProgram == 0) {
        LOGE("glCreateProgram failed. ret is %d", m_iProgram);
        release();
        return GL_CREATE_PROGRAM_ERROR;
    }

    glAttachShader(m_iProgram, m_iVertexShader);
    glAttachShader(m_iProgram, m_iFragmentShader);
    glLinkProgram(m_iProgram);
    GLint linkStatus = GL_FALSE;
    glGetProgramiv(m_iProgram, GL_LINK_STATUS, &linkStatus);
    if (linkStatus != GL_TRUE) {
        GLchar log[1024];
        GLsizei length;
        glGetProgramInfoLog(m_iProgram, 1024, &length, log);
        LOGE("glLinkProram failed. ret is %d, log: %s", linkStatus, log);
        release();
        return GL_LINK_PROGRAM_ERROR;
    }

    return READY;

}

void GLProgram::use() {
    if (m_iProgram != 0) {
        glUseProgram(m_iProgram);
    }
}

void GLProgram::unuse() {
    glUseProgram(0);
}

GLuint GLProgram::getProgramID() {
    return m_iProgram;
}

void GLProgram::release() {
    if (m_iVertexShader != 0) {
        glDeleteShader(m_iVertexShader);
        m_iVertexShader = 0;
    }
    if (m_iFragmentShader != 0) {
        glDeleteShader(m_iFragmentShader);
        m_iFragmentShader = 0;
    }
    if (m_iProgram != 0) {
        glDeleteProgram(m_iProgram);
        m_iProgram = 0;
    }
}